tlaplus/ghostferry.tla

```diff
@@ -407,33 +446,40 @@ PossibleRecords == Records \cup {NoRecordHere}
 }
 
  ***************************************************************************)
-\* BEGIN TRANSLATION
+\* BEGIN TRANSLATION - the hash of the PCal code: PCal-0fb2191b7ea3f2d9d2ddacddfb52ede1
 CONSTANT defaultInitValue
-VARIABLES CurrentMaxPrimaryKey, SourceTable, TargetTable, SourceBinlog, 
-          ApplicationReadonly, BinlogWriteQueue, TargetBinlogPos, 
-          BinlogStreamingStopRequested, pc, lastSuccessfulPK, currentRow, 
-          lastSuccessfulBinlogPos, currentBinlogEntry, oldRecord, newRecord, 
+VARIABLES MaxPrimaryKey, CurrentMaxPrimaryKey, LastSuccessfulPK, SourceTable,
+          TargetTable, SourceBinlog, ApplicationReadonly, BinlogWriteQueue,
+          TargetBinlogPos, BinlogStreamingStopRequested, pc, currentRow,
+          lastSuccessfulBinlogPos, currentBinlogEntry, oldRecord, newRecord,
           chosenPK
 
-vars == << CurrentMaxPrimaryKey, SourceTable, TargetTable, SourceBinlog, 
-           ApplicationReadonly, BinlogWriteQueue, TargetBinlogPos, 
-           BinlogStreamingStopRequested, pc, lastSuccessfulPK, currentRow, 
-           lastSuccessfulBinlogPos, currentBinlogEntry, oldRecord, newRecord, 
+vars == << MaxPrimaryKey, CurrentMaxPrimaryKey, LastSuccessfulPK, SourceTable,
+           TargetTable, SourceBinlog, ApplicationReadonly, BinlogWriteQueue,
+           TargetBinlogPos, BinlogStreamingStopRequested, pc, currentRow,
+           lastSuccessfulBinlogPos, currentBinlogEntry, oldRecord, newRecord,
            chosenPK >>
 
 ProcSet == {TableIterator} \cup {BinlogStreamer} \cup {BinlogWriter} \cup {Application} \cup {Ferry}
 
 Init == (* Global variables *)
+        /\ MaxPrimaryKey \in 1..TableCapacity
         /\ CurrentMaxPrimaryKey = MaxPrimaryKey
-        /\ SourceTable = InitialTable
-        /\ TargetTable = [k \in PrimaryKeys |-> NoRecordHere]
+        /\ LastSuccessfulPK \in 0..MaxPrimaryKey
+        /\ SourceTable \in                 {
+                               table \in PossibleTableStates : /\ \A k \in MaxPrimaryKey+1..TableCapacity : table[k] = NoRecordHere
+                                                               /\ ~(\A k \in 1..TableCapacity : table[k] = NoRecordHere)
+                           }
+        /\ TargetTable \in                 {
+                               table \in PossibleTableStates : /\ \A k \in 1..LastSuccessfulPK: table[k] = SourceTable[k]
```

Comment:
I think you are looking at an outdated diff. Not sure how this came to be, but the actual clause in the most up to date spec is:
```
    TargetTable \in {
        table \in PossibleTableStates : /\ \A k \in 1..ActualCopiedPK: table[k] = SourceTable[k]
                                        /\ \A k \in (ActualCopiedPK+1)..TableCapacity : table[k] = NoRecordHere
    },
```

where `ActualCopiedPK \in LastSuccessfulPK..MaxPrimaryKey,`

So basically for the target table, from 1..LastSuccessfulPK, it will definitely contain entries from the source table. But from LastSuccessfulPK..TableCapacity, it _could_ have entries from the source table, which mean ghostferry will recopy it.

Reply:
What happened is that this comment is attached to the generated TLA+ code, and I at one point had a generated TLA+ spec that is outdated.